bug(r2dbc): enable_iam_auth throws java.lang.ClassCastException

Due to a [limitation in R2DBC](https://github.com/r2dbc/r2dbc-spi/issues/206), extra options specified through the r2dbc url are automatically typed as `String`.  Therefore, setting the enable_iam_auth option in the r2dbc connection url (`r2dbc:gcp:postgres://{INSTANCE_CONNECTION_NAME}/{DATABASE_NAME}?ENABLE_IAM_AUTH=true`) results in the following error :

```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in module java.base of loader 'bootstrap')
	at com.google.cloud.sql.core.GcpConnectionFactoryProvider.createFactory(GcpConnectionFactoryProvider.java:103)
	at com.google.cloud.sql.core.GcpConnectionFactoryProvider.create(GcpConnectionFactoryProvider.java:96)
	at io.r2dbc.spi.ConnectionFactories.find(ConnectionFactories.java:112)
	at io.r2dbc.spi.ConnectionFactories.get(ConnectionFactories.java:142)
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder$OptionsCapableWrapper.buildAndWrap(ConnectionFactoryBuilder.java:203)
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder$PoolingAwareOptionsCapableWrapper.buildAndWrap(ConnectionFactoryBuilder.java:216)
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder.build(ConnectionFactoryBuilder.java:189)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations.createConnectionFactory(ConnectionFactoryConfigurations.java:61)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations$Pool.connectionFactory(ConnectionFactoryConfigurations.java:73)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 102 more
```
Perhaps doing something similar to [CoreSocketFactory#create](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/123889b52c70c6b8c94c4831211e677879207d19/core/src/main/java/com/google/cloud/sql/core/CoreSocketFactory.java#L226) where we call `Boolean.parseBoolean()` could help address this. Enabling IAM auth through the JDBC connection url (`jdbc:postgresql://google/{DATABASE_NAME}?socketFactory=com.google.cloud.sql.postgres.SocketFactory&cloudSqlInstance={INSTANCE_CONNECTION_NAME}?enableIamAuth=true&sslmode=disable` works as expected.

--- a/r2dbc/core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java
+++ b/r2dbc/core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java
@@ -100,13 +100,17 @@ public abstract class GcpConnectionFactoryProvider implements ConnectionFactoryP
       ConnectionFactoryOptions connectionFactoryOptions) {
     String connectionName = (String) connectionFactoryOptions.getRequiredValue(HOST);
     String socket = (String) connectionFactoryOptions.getValue(UNIX_SOCKET);
-    Boolean enableIamAuth = (Boolean) connectionFactoryOptions.getValue(ENABLE_IAM_AUTH);
+
+    Object iamAuthObj = connectionFactoryOptions.getValue(ENABLE_IAM_AUTH);
+    Boolean enableIamAuth = false;
+    if (iamAuthObj instanceof Boolean) {
+      enableIamAuth = (Boolean) iamAuthObj;
+    } else if (iamAuthObj instanceof String) {
+      enableIamAuth = Boolean.parseBoolean((String) iamAuthObj);
+    }
 
     Builder optionBuilder = createBuilder(connectionFactoryOptions);
 
-    if (enableIamAuth == null) {
-      enableIamAuth = false;
-    }
 
     // Precompute SSL Data to trigger the initial refresh to happen immediately,
     // and ensure enableIAMAuth is set correctly.

--- a/r2dbc/postgres/src/test/java/com/google/cloud/sql/core/R2dbcPostgresIamAuthIntegrationTests.java
+++ b/r2dbc/postgres/src/test/java/com/google/cloud/sql/core/R2dbcPostgresIamAuthIntegrationTests.java
@@ -147,4 +147,40 @@ public class R2dbcPostgresIamAuthIntegrationTests {
 
     assertThat(books).containsExactly("Book One", "Book Two");
   }
+
+  // This test verifies the "enable_iam_authn" parameter works when a URL is used
+  @Test
+  public void urlPooledConnectionTest() {
+    String url = String.format("r2dbc:gcp:postgres://%s:%s@%s/%s?ENABLE_IAM_AUTH=true", DB_USER, "password", CONNECTION_NAME, DB_NAME);
+    ConnectionFactory connectionPool = ConnectionFactories.get(url);
+    String insertStmt = String
+        .format("INSERT INTO  %s (ID, TITLE) VALUES ($1, $2)", this.tableName);
+    Mono.from(connectionPool.create())
+        .flatMapMany(
+            c ->
+                c.createStatement(insertStmt)
+                    .bind("$1", "book1")
+                    .bind("$2", "Book One")
+                    .add()
+                    .bind("$1", "book2")
+                    .bind("$2", "Book Two")
+                    .execute())
+        .flatMap(result -> result.map((row, rowMetadata) -> row.get(0)))
+        .blockLast();
+
+    String selectStmt = String.format("SELECT TITLE FROM %s ORDER BY ID", this.tableName);
+    List<String> books =
+        Mono.from(this.connectionPool.create())
+            .flatMapMany(
+                connection ->
+                    connection.createStatement(selectStmt).execute())
+            .flatMap(
+                result ->
+                    result.map(
+                        (r, meta) -> r.get("TITLE", String.class)))
+            .collectList()
+            .block();
+
+    assertThat(books).containsExactly("Book One", "Book Two");
+  }
 }

